File: 05_demography_survival.R
# 05 — Supplier demography and survival (canonical R port of
# 05_demography_survival.py; same outputs).
# Run: Rscript 05_demography_survival.R

source("theme_house.R", chdir = TRUE)

ERA_OF_YEAR <- function(y) ifelse(y <= 2019, "macri",
                                  ifelse(y <= 2023, "fernandez", "milei"))

adj <- read_parquet(file.path(DIR_PROC, "adjudicaciones_clean.parquet")) |>
  filter(es_nuevo, !is.na(provincia))

firm_years <- adj |>
  count(cuit, provincia, ejercicio, name = "n") |>
  mutate(region = region_of(provincia))

life <- firm_years |>
  group_by(cuit, provincia, region) |>
  summarise(entrada = min(ejercicio), salida = max(ejercicio),
            anios_activo = n_distinct(ejercicio), .groups = "drop") |>
  mutate(cohorte_era = ERA_OF_YEAR(entrada))

# ── Entry rates by region-year ───────────────────────────────────────────────
active <- firm_years |> group_by(region, ejercicio) |>
  summarise(activos = n_distinct(cuit), .groups = "drop")
entries <- life |> count(region, entrada, name = "entradas") |>
  rename(ejercicio = entrada)
ee <- active |> left_join(entries, by = c("region", "ejercicio")) |>
  mutate(tasa_entrada = entradas / activos)
write_tab(ee, "tab_entry_exit_region.csv")

# ── Cohort persistence across eras ───────────────────────────────────────────
active_by_era <- firm_years |>
  mutate(era = ERA_OF_YEAR(ejercicio)) |>
  distinct(cuit, era) |>
  mutate(v = TRUE) |>
  pivot_wider(names_from = era, values_from = v, values_fill = FALSE)

surv_of <- function(cuits, target) {
  mean(active_by_era[[target]][match(cuits, active_by_era$cuit)], na.rm = FALSE)
}

pairs <- list(c("macri", "fernandez"), c("macri", "milei"),
              c("fernandez", "milei"))
cs <- bind_rows(lapply(c("NEA", "Metro", "Resto"), function(reg) {
  bind_rows(lapply(pairs, function(pr) {
    cuits <- life$cuit[life$region == reg & life$cohorte_era == pr[1]]
    if (!length(cuits)) return(NULL)
    tibble(region = reg, cohorte = pr[1], sobrevive_en = pr[2],
           n_cohorte = length(cuits),
           tasa_supervivencia = round(surv_of(cuits, pr[2]), 3))
  }))
}))
cs <- bind_rows(cs, bind_rows(lapply(sort(NEA), function(prov) {
  cuits <- life$cuit[life$provincia == prov & life$cohorte_era == "macri"]
  if (!length(cuits)) return(NULL)
  tibble(region = prov, cohorte = "macri", sobrevive_en = "milei",
         n_cohorte = length(cuits),
         tasa_supervivencia = round(surv_of(cuits, "milei"), 3))
})))
write_tab(cs, "tab_cohort_survival.csv")

# ── Discrete survival curves (entrants <= 2021) ──────────────────────────────
curves <- bind_rows(lapply(REGION_LEVELS, function(reg) {
  sub <- life |> filter(region == reg, entrada <= 2021)
  bind_rows(lapply(0:5, function(k) {
    at_risk <- sub |> filter(entrada + k <= 2026)
    tibble(region = reg, k = k, n = nrow(sub),
           surv = mean(at_risk$salida >= at_risk$entrada + k))
  }))
})) |>
  mutate(region = factor(region, levels = REGION_LEVELS))

p <- ggplot(curves, aes(k, surv, colour = region)) +
  geom_line() + geom_point() +
  scale_colour_manual(values = REGION_COLS) +
  labs(x = "Años desde la primera adjudicación",
       y = "P(sigue adjudicando en el año k o después)",
       title = "Supervivencia de proveedores (entrantes 2016-2021)") +
  theme_house()
save_fig(p, "fig_survival_by_region.png")

cat("\nPersistencia de cohortes:\n"); print(cs, n = 30)
cat("\nEntradas por región (2022+):\n")
print(ee |> filter(ejercicio >= 2022) |>
        mutate(tasa_entrada = round(tasa_entrada, 3)), n = 20)

File: 05_demography_survival.py
"""
05 — Supplier demography and survival across political transitions
==================================================================
Project: Who Sells to the State (2026_2)

Entry (first award), exit (last award), tenure; annual entry rates by
region; cohort persistence across presidencies (P(era-X entrant still
winning under era Y)) by province/region; discrete survival curves
(pandas, no lifelines).

Censoring: exit is right-censored at 2025+ (a supplier whose last award
is 2025-2026 may still be active); survival curves use awards observed
through Aug-2026.

Input:  data/processed/adjudicaciones_clean.parquet
Output: tables/tab_cohort_survival.csv
        tables/tab_entry_exit_region.csv
        figures/fig_survival_by_region.png

Usage:  python 05_demography_survival.py
"""
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

PROJECT = Path(__file__).parent
PROC = PROJECT / "data" / "processed"
TAB = PROJECT / "tables"
FIG = PROJECT / "figures"

NEA = {"Misiones", "Corrientes", "Chaco", "Formosa"}
METRO = {"CABA", "Buenos Aires"}
ERA_OF_YEAR = {**{y: "macri" for y in range(2016, 2020)},
               **{y: "fernandez" for y in range(2020, 2024)},
               **{y: "milei" for y in range(2024, 2027)}}

adj = pd.read_parquet(PROC / "adjudicaciones_clean.parquet")
adj = adj[adj["es_nuevo"] & adj["provincia"].notna()]

firm_years = (adj.groupby(["cuit", "provincia", "ejercicio"])
              .size().rename("n").reset_index())
firm_years["region"] = np.select(
    [firm_years["provincia"].isin(NEA), firm_years["provincia"].isin(METRO)],
    ["NEA", "Metro"], default="Resto")

life = (firm_years.groupby(["cuit", "provincia", "region"])
        .agg(entrada=("ejercicio", "min"), salida=("ejercicio", "max"),
             anios_activo=("ejercicio", "nunique"))
        .reset_index())
life["cohorte_era"] = life["entrada"].map(ERA_OF_YEAR)

# ── Entry/exit rates by region-year ──────────────────────────────────────────
active = firm_years.groupby(["region", "ejercicio"])["cuit"].nunique().rename("activos")
entries = life.groupby(["region", "entrada"])["cuit"].size().rename("entradas")
ee = pd.concat([active, entries.rename_axis(["region", "ejercicio"])], axis=1)
ee["tasa_entrada"] = ee["entradas"] / ee["activos"]
ee.reset_index().to_csv(TAB / "tab_entry_exit_region.csv", index=False)

# ── Cohort persistence across eras ───────────────────────────────────────────
active_by_era = (firm_years.assign(era=firm_years["ejercicio"].map(ERA_OF_YEAR))
                 .groupby(["cuit", "era"]).size().unstack().notna())
rows = []
for scope, sel in [("NEA", life["region"] == "NEA"),
                   ("Metro", life["region"] == "Metro"),
                   ("Resto", life["region"] == "Resto")]:
    for cohorte, target in [("macri", "fernandez"), ("macri", "milei"),
                            ("fernandez", "milei")]:
        cuits = life[sel & (life["cohorte_era"] == cohorte)]["cuit"]
        if len(cuits) == 0:
            continue
        surv = active_by_era.reindex(cuits)[target].fillna(False).mean()
        rows.append({"region": scope, "cohorte": cohorte, "sobrevive_en": target,
                     "n_cohorte": len(cuits), "tasa_supervivencia": round(surv, 3)})
cs = pd.DataFrame(rows)
# per NEA province detail, macri -> milei
for prov in sorted(NEA):
    cuits = life[(life["provincia"] == prov) & (life["cohorte_era"] == "macri")]["cuit"]
    if len(cuits):
        surv = active_by_era.reindex(cuits)["milei"].fillna(False).mean()
        cs = pd.concat([cs, pd.DataFrame([{
            "region": prov, "cohorte": "macri", "sobrevive_en": "milei",
            "n_cohorte": len(cuits), "tasa_supervivencia": round(surv, 3)}])])
cs.to_csv(TAB / "tab_cohort_survival.csv", index=False)

# ── Discrete survival curves (age = years since entry, entrants <= 2021) ─────
fig, ax = plt.subplots(figsize=(8, 4.5))
for reg, color in [("NEA", "#d62728"), ("Metro", "#555555"), ("Resto", "#1f77b4")]:
    sub = life[(life["region"] == reg) & (life["entrada"] <= 2021)]
    ages = np.arange(0, 6)
    surv = []
    for k in ages:
        at_risk = sub[sub["entrada"] + k <= 2026]
        alive = (at_risk["salida"] >= at_risk["entrada"] + k).mean()
        surv.append(alive)
    ax.plot(ages, surv, "o-", color=color, label=f"{reg} (n={len(sub)})")
ax.set_xlabel("Años desde la primera adjudicación")
ax.set_ylabel("P(sigue adjudicando en el año k o después)")
ax.set_title("Supervivencia de proveedores en el mercado nacional (entrantes 2016-2021)")
ax.legend()
fig.tight_layout()
fig.savefig(FIG / "fig_survival_by_region.png", dpi=200)
plt.close(fig)

print("Persistencia de cohortes entre presidencias:")
print(cs.to_string(index=False))
print("\nEntradas y activos por región (últimos años):")
print(ee.reset_index().query("ejercicio >= 2022").round(3).to_string(index=False))
print("OK -> tables/, figures/")
